k9mail/src/test/java/com/fsck/k9/controller/MessagingControllerTest.java

```diff
@@ -0,0 +1,379 @@
+
+        controller.synchronizeMailboxSynchronous(account, FOLDER_NAME, listener, remoteFolder);
+
+        verify(remoteFolder, atLeast(2)).fetch(any(List.class), fetchProfileCaptor.capture(), any(MessageRetrievalListener.class));
+        assertEquals(1, fetchProfileCaptor.getAllValues().get(1).size());
+        assertTrue(fetchProfileCaptor.getAllValues().get(1).contains(FetchProfile.Item.BODY));
+    }
+
+    @Test
+    public void synchronizeMailboxSynchronous_withUnsyncedNewSmallMessage_shouldFetchStructureAndLimitedBodyOfLargeMessage() throws Exception {
+        final Message largeMessage = buildLargeNewMessage();
+
+        messageCountInRemoteFolder(1);
+        hasUnsychedRemoteMessage();
+        when(remoteFolder.supportsFetchingFlags()).thenReturn(false);
+        respondToFetchEnvelopesWithMessage(largeMessage);
+
+        controller.synchronizeMailboxSynchronous(account, FOLDER_NAME, listener, remoteFolder);
+
+        //TODO: Don't bother fetching messages of a size we don't have
+        verify(remoteFolder, atLeast(4)).fetch(any(List.class), fetchProfileCaptor.capture(), any(MessageRetrievalListener.class));
+        assertEquals(1, fetchProfileCaptor.getAllValues().get(2).size());
+        assertEquals(FetchProfile.Item.STRUCTURE, fetchProfileCaptor.getAllValues().get(2).get(0));
+        assertEquals(1, fetchProfileCaptor.getAllValues().get(3).size());
+        assertEquals(FetchProfile.Item.BODY_SANE, fetchProfileCaptor.getAllValues().get(3).get(0));
+    }
+
+
+
+    private void respondToFetchEnvelopesWithMessage(final Message message) throws MessagingException {
+        doAnswer(new Answer() {
+            @Override
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+                FetchProfile fp = (FetchProfile) invocation.getArguments()[1];
+                if(invocation.getArguments()[2] != null) {
+                    MessageRetrievalListener l = (MessageRetrievalListener) invocation.getArguments()[2];
+                    if (fp.contains(FetchProfile.Item.ENVELOPE)) {
+                        l.messageStarted("UID", 1, 1);
+                        l.messageFinished(message, 1, 1);
+                        l.messagesFinished(1);
+                    }
+                }
+                return null;
+            }
+        }).when(remoteFolder).fetch(any(List.class), any(FetchProfile.class), any(MessageRetrievalListener.class));
+
+    }
+
+    private Message buildSmallNewMessage() {
+        Message message = mock(Message.class);
+        when(message.olderThan(any(Date.class))).thenReturn(false);
+        when(message.getSize()).thenReturn(MAXIMUM_SMALL_MESSAGE_SIZE);
+        return message;
+    }
+
+    private Message buildLargeNewMessage() {
+        Message message = mock(Message.class);
+        when(message.olderThan(any(Date.class))).thenReturn(false);
+        when(message.getSize()).thenReturn(MAXIMUM_SMALL_MESSAGE_SIZE+1);
```

Comment:
Code style issue: spaces missing in expression; should be `MAXIMUM_SMALL_MESSAGE_SIZE + 1`